lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,540 @@
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+        Self { interval }
+    }
+}
+
+impl Timer for BatcherTimer {
+    fn poll_elapsed(&mut self, cx: &mut Context) -> Poll<bool> {
+        match self.interval.poll_tick(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(_) => Poll::Ready(true),
+        }
+    }
+}
+
+/// A batch for use by `Batcher`
+///
+/// This structure is a private implementation detail that simplifies the
+/// implementation of `Batcher`. It is the actual store of items that come
+/// through the stream manipulated by `Batcher` plus limit information to signal
+/// when the `Batch` is full.
+struct Batch<I> {
+    /// The total number of `I` bytes stored, does not any overhead in this
+    /// structure.
+    allocated_bytes: usize,
+    /// The maximum number of elements allowed in this structure.
+    element_limit: usize,
+    /// The maximum number of allocated bytes (not including overhead) allowed
+    /// in this structure.
+    allocation_limit: usize,
+    /// The store of `I` elements.
+    elements: Vec<I>,
+}
+
+impl<I> ByteSizeOf for Batch<I> {
+    fn allocated_bytes(&self) -> usize {
+        self.allocated_bytes
+    }
+}
+
+impl<I> Batch<I>
+where
+    I: ByteSizeOf,
+{
+    /// Create a new Batch instance
+    ///
+    /// Creates a new batch instance with specific element and allocation
+    /// limits. The element limit is a maximum cap on the number of `I`
+    /// instances. The allocation limit is a soft-max on the number of allocated
+    /// bytes stored in this batch, not taking into account overhead from this
+    /// structure itself. Caller is responsible for ensuring that `I` will fit
+    /// inside the allocation limit.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the allocation limit will not store at least
+    /// 1 instance of `I`, as measured by `mem::size_of`.
+    fn new(element_limit: usize, allocation_limit: usize) -> Self {
+        assert!(allocation_limit >= mem::size_of::<I>());
+        Self {
+            allocated_bytes: 0,
+            element_limit,
+            allocation_limit,
+            elements: Vec::with_capacity(element_limit),
+        }
+    }
+
+    /// Unconditionally insert an element into the batch
+    ///
+    /// This function is similar to `push` except that the caller does not need
+    /// to call `has_space` prior to calling this and it will never
+    /// panic. Intended to be used only when insertion must not fail.
+    fn with(mut self, value: I) -> Self {
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+        self
+    }
+
+    /// Decompose the batch
+    ///
+    /// Called by the user when they want to get at the internal store of
+    /// items. Returns a tuple, the first element being the allocated size of
+    /// stored items and the second the store of items.
+    fn destruct(self) -> (usize, Vec<I>) {
```

Comment:
`into_X` (ie `into_inner`) is a more common naming convention for something like this.

Reply:
I'm open to suggestions but couldn't think of a good `into_X` name for this, other than maybe `into_allocated_bytes_and_inner` which is, uh, big.

Reply:
This is also something I've been staring at for a while, so it's challenging to see not from up-close.

Reply:
It's _almost_ a `into_parts`, so that might be close enough. Maybe just `into_buffer`? I don't have a strong feeling for the specific name, just that `destruct` is a side effect of taking ownership of `self`, and doesn't appear to be the actual purpose.

Reply:
Resolved in af6f9563f